Challenge-4-dex First Draft

mertcanciy opened the pull request.

rin-st commented on README.md:
```diff
@@ -1,11 +1,13 @@
 Deploy your contracts to a testnet then build and upload your app to a public web server. Submit the url on [SpeedRunEthereum.com](https://speedrunethereum.com)!
 
-💬 Meet other builders working on this challenge and get help in the {challengeTelegramLink}
+There is also a 🎥 [Youtube video](https://www.youtube.com/watch?v=eP5w6Ger1EQ&t=364s&ab_channel=SimplyExplained) that may help you understand the concepts covered within this challenge too:
```
change it to https://www.youtube.com/watch?v=eP5w6Ger1EQ , so it will not point to the end of the video. ab_channel also not needed

rin-st commented on README.md:
```diff
@@ -33,19 +35,393 @@ yarn chain
+
+👩‍💻 Rerun yarn deploy whenever you want to deploy new contracts to the frontend (run yarn deploy --reset for a completely fresh deploy if you have made no contract changes).
+
+Balloons.sol is just an example ERC20 contract that mints 1000 $BAL to whatever address deploys it. DEX.sol is what we will build in this challenge and you can see it starts with a SafeMath library to help us prevent overflows and underflows and also tracks a token (ERC20 interface) that we set in the constructor (on deploy).
```
backticks for balloons.sol and dex.sol. Also probably we need to name it `packages/hardhat/contracts/Baloons.sol` like it's in other challenges

rin-st commented on README.md:
```diff
@@ -33,19 +35,393 @@ yarn chain
+
+![2023-07-13 (2)](https://github.com/mertcanciy/se-2-challenges/assets/59885513/1b6e8f59-c549-40b8-982e-e0bec0239117)
+
+
```
there's no this block of text, why?

<img width="1098" alt="image" src="https://github.com/scaffold-eth/se-2-challenges/assets/25638585/9ebb3350-d6d4-4fb1-9d27-5a8a501b48d9">

rin-st commented on README.md:
```diff
@@ -33,19 +35,393 @@ yarn chain
+
+Balloons.sol is just an example ERC20 contract that mints 1000 $BAL to whatever address deploys it. DEX.sol is what we will build in this challenge and you can see it starts with a SafeMath library to help us prevent overflows and underflows and also tracks a token (ERC20 interface) that we set in the constructor (on deploy).
+
+Below is what your front-end will look like with no implementation code within your smart contracts yet. The buttons will likely break because there are no functions tied to them yet!
```
in original readme there is `>` before this paragraph

rin-st commented on README.md:
```diff
@@ -33,19 +35,393 @@ yarn chain
+
+Below is what your front-end will look like with no implementation code within your smart contracts yet. The buttons will likely break because there are no functions tied to them yet!
+
+⭐️ Also note that there is no curve until you uncomment the specific lines of code at the end of `hardhat/deploy/00_deploy_your_contract.ts`.
```
and `> **⭐️ Also...` here

rin-st replied:
probably you copy-pasted text, not `README.md` code, recheck it please

rin-st commented on README.md:
````diff
@@ -33,19 +35,393 @@ yarn chain
+
+We can see that the DEX starts empty. We want to be able to call `init()` to start it off with liquidity, but we don’t have any funds or tokens yet. Add some ETH to your local account using the faucet and then find the `00_deploy_your_contract.ts` file. Find and uncomment the line below and add your front-end address:
+
+```
````
comment this block in deployment file

rin-st commented on packages/hardhat/deploy/00_deploy_your_contract.ts:
```diff
@@ -21,22 +21,54 @@ const deployYourContract: DeployFunction = async function (hre: HardhatRuntimeEn
+
+  const dex = await hre.ethers.getContract("DEX", deployer);
+
+    // paste in your front-end address here to get 10 balloons on deploy:
```
this block should also be commented https://github.com/scaffold-eth/se-2-challenges/pull/15/files#r1269525687

rin-st commented on README.md:
```diff
@@ -33,19 +35,393 @@ yarn chain
+        uint256 xInput,
+        uint256 xReserves,
+        uint256 yReserves
+    ) public view returns (uint256 yOutput) {
```
we can change it to `pure` instead of `view`

rin-st commented on README.md:
```diff
@@ -33,19 +35,393 @@ yarn chain
+
+When we call `init()` we passed in ETH and $BAL tokens at a ratio of 1:1. As the reserves of one asset changes, the other asset must also change inversely in order to maintain the constant product formula (invariant k).
+
+Now, try to edit your DEX.sol smart contract and bring in a price function!
```
`DEX.sol`

rin-st commented on README.md:
```diff
@@ -33,19 +35,393 @@ yarn chain
+
+---
+
+## Checkpoint 7: 💾 Deploy it! 🛰
```
Block removed, also tests file didn't copied to the repo
<img width="1134" alt="image" src="https://github.com/scaffold-eth/se-2-challenges/assets/25638585/c91e7153-1cae-4170-8951-510328bcb555">

rin-st commented on packages/nextjs/pages/index.tsx:
```diff
@@ -1,24 +1,177 @@
+            </div>
+            <div className="relative flex py-5 items-center">
+                <div className="flex-grow border-t border-gray-400"></div>
+                <span className="flex-shrink mx-4 text-gray-400">Liquidity(None)</span>
```
It's not always none, check original https://github.com/scaffold-eth/scaffold-eth-challenges/blob/challenge-4-dex/packages/react-app/src/components/DEX.jsx#L110

rin-st commented:
Hey @mertcanciy ! Great job! Added comments mostly regarding readme.

Regarding frontend. For more consistency, let's try to do it similar to challenges 1 and 2. Something like this
<img width="748" alt="image" src="https://github.com/scaffold-eth/se-2-challenges/assets/25638585/43385b9b-2a6a-49b6-9ed6-1de38410737c">
but with two blocks: Dex contract and Baloons contract.

rin-st commented on packages/nextjs/pages/index.tsx:
```diff
@@ -1,24 +1,177 @@
+                {isLoading ? (
+                  <span>Loading...</span>
+                ) : (
+                  <span>Balloons: {ethers.utils.formatEther(DEXBalloonBalance?.toString() || 0)}</span>
```
<img width="420" alt="image" src="https://github.com/scaffold-eth/se-2-challenges/assets/25638585/f82909cd-3a7b-44f6-969e-f6874b58442e">
where is this icons?)

rin-st replied:
and balloons balance should also be `.toFixed(4)`

rin-st commented on packages/nextjs/pages/index.tsx:
```diff
@@ -1,24 +1,177 @@
+
+              <div className="flex mb-4">
+                
+                <span>Deposit  <EtherInput value={depositAmount} onChange={(value) => setDepositAmount(value)}/></span>
```
IntegerInput

rin-st commented on packages/nextjs/pages/index.tsx:
```diff
@@ -1,24 +1,177 @@
+              </div>
+
+              <div className="flex">
+                <span>Withdraw  <EtherInput value={withdrawAmount} onChange={(value) => setWithdrawAmount(value)}/></span>
```
IntegerInput

rin-st commented on packages/nextjs/pages/index.tsx:
```diff
@@ -1,24 +1,177 @@
+
+            <div className="border rounded-3xl px-4 py-3 grid grid-cols-2">
+              <div className="flex flex-col gap-4 mb-4">
+                <span className="flex flex-col">Approve  <input className="pl-4 py-4 px-8 text-accent bg-transparent border border-blue-100 rounded-2xl h-[2.2rem] min-h-[2.2rem]" value={approveSpender} onChange={(event) => setApproveSpender(event.target.value)} placeholder="Address Spender"/></span>
```
AddressInput

rin-st commented on packages/nextjs/pages/index.tsx:
```diff
@@ -1,24 +1,177 @@
+            <div className="border rounded-3xl px-4 py-3 grid grid-cols-2">
+              <div className="flex flex-col gap-4 mb-4">
+                <span className="flex flex-col">Approve  <input className="pl-4 py-4 px-8 text-accent bg-transparent border border-blue-100 rounded-2xl h-[2.2rem] min-h-[2.2rem]" value={approveSpender} onChange={(event) => setApproveSpender(event.target.value)} placeholder="Address Spender"/></span>
+                <EtherInput value={approveAmount} onChange={(value) => setApproveAmount(value)} placeholder="uint256 Amount"/>
```
IntegerInput

rin-st commented on packages/nextjs/pages/index.tsx:
```diff
@@ -1,24 +1,177 @@
+
+            <div className="border rounded-3xl px-4 py-3 mt-6">
+              <div className="flex mb-4">
+                <span className="flex flex-col mt-2">balanceOf <input className="pl-4 py-4 px-8 text-accent bg-transparent border border-blue-100 rounded-2xl h-[2.2rem] min-h-[2.2rem]" value={accountBalanceOf} onChange={(event) => setAccountBalanceOf(event.target.value)} placeholder="address Account"/></span>                
```
AddressInput

rin-st commented:
Other than that there are some places where you need to center/align things, but we can think about it after https://github.com/scaffold-eth/se-2-challenges/pull/15#issuecomment-1644159501

Pabl0cks commented:
Just adding a point so you can review it for your next version @mertcanciy. I've realized you have added all the logic in the `index.tsx` (`Home` tab)

You can check how we do it in other challenges or check this [💡 Guide and Hints to create New Challenges](https://github.com/scaffold-eth/se-2-challenges#-guide-and-hints-to-create-new-challenges).

We leave the `Home` tab for a banner image + the Challenge description, and add the Challenge logic in different tabs. Here are Challenge 2 tabs for example:

![challenge-2](https://user-images.githubusercontent.com/55535804/254721435-7669cc68-e942-4630-95c8-91cd21af5ba0.png)

Good job with the PR 🙌

mertcanciy commented:
Hey @rin-st @Pabl0cks ! Thank you so much for your great comments. I tried to fix all of the issues you mentioned. There are 2 issues I couldn't handle:
1- I couldn't understand how I can change the photos' URL. https://github.com/scaffold-eth/se-2-challenges/pull/15#discussion_r1269451371

2- I couldn't handle the IntegerInput component. Now, when the user enter the values to input fields, all structure works but I could not handle the IntegerInput's multiplication logic. If you have time, can you check it out?  

And also, I put the challenge page to DEX.tsx as @Pabl0cks mentioned and in the index.tsx, there is only the description of challenge.

Waiting for your reviews!

rin-st commented:
hey, @mertcanciy awesome!
1. You can't change it from your repo. Just add todo and we'll change it later
2. Regarding IntegerInput, it's behavior will be changed after updating sre-2 to viem, and we'll need to add props like `hideMultiplier` and fix bugs if we'll find some. For now, you can check challenge-2-token-vendor branch and see how it's implemented there. Also, copy utility functions if needed

rin-st commented on README.md:
```diff
@@ -33,19 +35,402 @@ yarn chain
+
+👩‍💻 Rerun `yarn deploy` whenever you want to deploy new contracts to the frontend (run `yarn deploy --reset` for a completely fresh deploy if you have made no contract changes).
+
+`packages/hardhat/contracts/Baloons.sol` is just an example ERC20 contract that mints 1000 $BAL to whatever address deploys it. `packages/hardhat/contracts/DEX.sol` is what we will build in this challenge and you can see it starts with a SafeMath library to help us prevent overflows and underflows and also tracks a token (ERC20 interface) that we set in the constructor (on deploy).
```
Balloons

rin-st commented:
UI became much better! Let's try to improve it a little bit
- [ ] Make titles centered and similar to https://github.com/scaffold-eth/se-2-challenges/pull/15#issuecomment-1644159501 Staker contract.
- [ ] Make inputs full-width or centered with the same width
- [ ] Make "Liquidity" text bigger and remove the line
- [ ] Align chart to the top or try to make it `position: sticky` if possible

<img width="712" alt="Снимок экрана 2023-07-23 в 17 22 45" src="https://github.com/scaffold-eth/se-2-challenges/assets/25638585/3cee0a3a-a0fd-45a7-870d-ccb8111aa585">

<img width="698" alt="Снимок экрана 2023-07-23 в 17 31 21" src="https://github.com/scaffold-eth/se-2-challenges/assets/25638585/00da3d5e-5441-4d3c-8aee-a434039e31ba">

rin-st commented:
And hide multiplier for integer input. You can do it like here https://github.com/scaffold-eth/se-2-challenges/pull/23/files#diff-fa1b7c802d3da5c87e09caf71184241d6682773c9fe9fac6e6d104201fe21d64 (hideSuffix prop)

mertcanciy commented:
hey @rin-st ! I think I almost handled all the comments you mentioned if I did not missed any of them but the only one I couldn't make is that I could not aligned the cart to the top :/

Pabl0cks commented:
> hey @rin-st ! I think I almost handled all the comments you mentioned if I did not missed any of them but the only one I couldn't make is that I could not aligned the cart to the top :/

Nice job @mertcanciy, I think tomorrow morning I will be able to fully test the challenge and give you my feedback.

I'm not a tailwind wizard, so may not be the perfect approach. But here it is:
- I think you can align it to the top changing `items-center` for `items-start` in the parent `<div>`, just after the `<h1>`. Could be something like this: `<div className="flex flex-col flex-grow items-start pt-10 grid grid-cols-2 content-start">`
- If you want to also set it sticky, can add this to the  `<div>` that contains the `<Curve>` component. Something like this: 
```
<div className="sticky top-0 px-5 mx-auto">
          <Curve
```

mertcanciy commented:
> 

It worked, thanks!

Pabl0cks commented:
Did a bit of testing this morning, comparing old challenge deploy vs migrated challenge, here are the main issues:

- [ ] We are missing the Events tab we had in the old challenge (was called Eventlist). It was showing all the EthToTokenSwap Events, TokenToEthSwap Events, LiquidityProvided Events and LiquidityRemoved Events. You can check Challenge 2 for example
- [ ] I can't see the balance of my wallet, neither for balloons nor liquidity provided. In the old challenge we had it below logout, not sure if it's the best place, we can think about it.
![image](https://github.com/scaffold-eth/se-2-challenges/assets/55535804/be0b73eb-22f1-4916-bcb0-de72ebf25014)
- [ ] Minor issue. When you already have a value in `ethToToken` or `tokenToEth` and put a value in the other field, it used to refresh the graph with the new conversion, but in the migrated one it doesn't.

Pabl0cks commented:
I've just pushed some small tweaks, here is a couple of issues I found this morning:
- [ ] If you try to exchange `tokenToETH` without enough allowance, now is throwing a 'ERC20: insufficient allowance' error, but in pre-migration challenge it was sending a first transaction to increase the allowance, and a second transaction for the token exchange when the first one was finished.
- [ ] Deposit liquidity failed if you did not have enough allowance in pre-migration version, same as current version. Should we leave it like that, or should we send a previous transaction to increase allowance?
- [ ] When running yarn hardhat:test I had to comment a couple of lines of `challenge-4.ts` to make it work. I'm not sure if it's enough with that or we should change any other code from the test. Those are the lines that gave me problems:
    - //import { solidity } from "ethereum-waffle";
    - //use(solidity);

cc @rin-st for technical opinion on those items 🧐

carletex commented:
> When running yarn hardhat:test I had to comment a couple of lines of challenge-4.ts to make it work. I'm not sure if it's enough with that or we should change any other code from the test. Those are the lines that gave me problems:
//import { solidity } from "ethereum-waffle";
//use(solidity);

Yep, this is not needed anymore (it was mandatory on earlier versions of hardhat).

Also, let's keep consistency with the naming of the test file in other challenges. It should be `Challenge4.ts`

mertcanciy commented:
> Did a bit of testing this morning, comparing old challenge deploy vs migrated challenge, here are the main issues:
> 
> * [ ]  We are missing the Events tab we had in the old challenge (was called Eventlist). It was showing all the EthToTokenSwap Events, TokenToEthSwap Events, LiquidityProvided Events and LiquidityRemoved Events. You can check Challenge 2 for example
> * [ ]  I can't see the balance of my wallet, neither for balloons nor liquidity provided. In the old challenge we had it below logout, not sure if it's the best place, we can think about it.
>   ![image](https://user-images.githubusercontent.com/55535804/256055292-be0b73eb-22f1-4916-bcb0-de72ebf25014.png)
> * [ ]  Minor issue. When you already have a value in `ethToToken` or `tokenToEth` and put a value in the other field, it used to refresh the graph with the new conversion, but in the migrated one it doesn't.

In token vendor challenge's Vendor contract, variable names of the event were specified and already commented in the beginning of the challenge
![image](https://github.com/scaffold-eth/se-2-challenges/assets/59885513/40dffdba-4808-415c-9b38-6e1b4057433e)
but here, the inside of events are empty. 
![image](https://github.com/scaffold-eth/se-2-challenges/assets/59885513/22634215-6b3b-430e-8546-245c00a9da03)
What can we do for it? 
If we don't specify the variables of events, how can we use it in Events.tsx? I'm a little bit confused

mertcanciy commented:
Hey @rin-st @Pabl0cks ! I added the variables that are inside of events to the Dex.sol file
![image](https://github.com/scaffold-eth/se-2-challenges/assets/59885513/0a966fd1-6449-4c9c-b57e-9b89c4c1e395)
Now, events are working in Events tab. 
It looks like this:
![image](https://github.com/scaffold-eth/se-2-challenges/assets/59885513/859ddbdd-c8b1-4770-aa03-2490955ca36b)


- I could not understand the logic of sending extra transaction when the user get the insufficient allowance error. I checked the transaction and yeah, it's true but I could not understand what's the point of that transaction.
https://github.com/scaffold-eth/se-2-challenges/pull/15#issuecomment-1651735462

- I removed the codes that causes problem in test file and changed its name to Challenge4.ts as @carletex said.
- I added the user's Balloons and Liquidity values to the right-top corner just like the old one.

I really made you tired, sorry for that btw :)

rin-st commented:
Good job @mertcanciy and @Pabl0cks ! Now it looks much better!

>  - [ ] If you try to exchange tokenToETH without enough allowance, now is throwing a 'ERC20: insufficient allowance' error, but in pre-migration challenge it was sending a first transaction to increase the allowance, and a second transaction for the token exchange when the first one was finished.
> - [ ] Deposit liquidity failed if you did not have enough allowance in pre-migration version, same as current version. Should we leave it like that, or should we send a previous transaction to increase allowance?

I think in both cases user should update allowances manually, we have inputs for it. I think it will help him to understand how it works and why, and also he will not think that some contract can spend his tokens without permission. @carletex what do you think?

carletex commented:
> I think in both cases user should update allowances manually, we have inputs for it. I think it will help him to understand how it works and why, and also he will not think that some contract can spend his tokens without permission. @carletex what do you think?

I agree. For me, this makes more sense, since it teaches the user how the approval/allowance pattern works.

But also asking @ZakGriffith here, since he has more context on this.

ZakGriffith commented:
> 

Yes, I agree as well.  At this point, most of the people running this challenge will have completed challenge 2 and learned a bit about the approve pattern, but having them manually approve the tokens drives the point home.  I haven't yet ran through the ReadMe of this challenge in SE2, but it may be appropriate to have something explaining the approve and allowance functions.

Pabl0cks commented:
Nice idea, I'm reviewing the README to ensure consistency with the previous challenges, can add the approve and allowance functions.

Pabl0cks commented:
Nice fixes @mertcanciy! Did some small tweaks to Events, but it was working great.

> I added the variables that are inside of events to the Dex.sol file

We'll have to update emit events from README Solution Code snippets to match your variables, I can take that change if you want when I finish reviewing README.

> Minor issue. When you already have a value in ethToToken or tokenToEth and put a value in the other field, it used to refresh the graph with the new conversion, but in the migrated one it doesn't.

Added a fix for that, maybe is not the cleanest solution, if there is a better way to fix it, feel free to undo my change 🙌

mertcanciy commented:
Thanks for your comments @carletex @ZakGriffith ! And also thanks for your commits @Pabl0cks @rin-st ! 

* I changed the variables of emits in Readme's solution snippets to match the variables in events. 
What I'd like to ask you that in Readme, there is also an explanation for the error and the solution of approve/allowance pattern.
![image](https://github.com/scaffold-eth/se-2-challenges/assets/59885513/a5637374-5f86-44f7-83cf-7606a2ff0b2e)

Is this explanation not enough? If not, I can also add extra explanation to the part I sent as photo here about how approve and allowance functions works or maybe how they work.

Pabl0cks commented:
> I changed the variables of emits in Readme's solution snippets to match the variables in events.
What I'd like to ask you that in Readme, there is also an explanation for the error and the solution of approve/allowance pattern.

Great! I was about to do it, so just pulled your changes and then pushed my tweaks on Readme texts and updated the screenshots pointing to your repo. I've also changed them to light mode to match the rest of the screenshots.

> Is this explanation not enough? If not, I can also add extra explanation to the part I sent as photo here about how approve and allowance functions work.

I've added some explanation about that, can review it and adapt it if feels not clear yet.

@ZakGriffith could you take a look at the Readme and try this challenge when you have some spare time? Thanks!  

I feel we are getting closer to mark this Challenge as completed, looking much more solid now 🙌

mertcanciy commented:
> > I changed the variables of emits in Readme's solution snippets to match the variables in events.
> > What I'd like to ask you that in Readme, there is also an explanation for the error and the solution of approve/allowance pattern.
> 
> Great! I was about to do it, so just pulled your changes and then pushed my tweaks on Readme texts and updated the screenshots pointing to your repo. I've also changed them to light mode to match the rest of the screenshots.
> 
> > Is this explanation not enough? If not, I can also add extra explanation to the part I sent as photo here about how approve and allowance functions work.
> 
> I've added some explanation about that, can review it and adapt it if feels not clear yet.
> 
> @ZakGriffith could you take a look at the Readme and try this challenge when you have some spare time? Thanks!
> 
> I feel we are getting closer to mark this Challenge as completed, looking much more solid now 🙌

Thanks for the changes! Yeah I think the challenge is almost done too. We can wait for the feedbacks now!

ZakGriffith commented:
Running this today and making notes of my findings. This feels far enough along to merge, then continue with new Issues/PRs.  Thoughts?

mertcanciy commented:
> Running this today and making notes of my findings. This feels far enough along to merge, then continue with new Issues/PRs. Thoughts?

I aggree with you ser! We can keep improving and fixing the challenge until the end of the migration for all the challenges

ZakGriffith commented:
Prefect, merging this beast into challenge-4-dex.  If any Issues still exist, let's create new to track them a little easier.  Good work on this!
